proto/regen/ecocredit/v1alpha1/events.proto

```diff
@@ -103,3 +105,95 @@ message EventCancel {
   // amount is the decimal number of credits that have been cancelled.
   string amount = 3;
 }
+
+// EventSell is an event emitted when a sell order is created.
+message EventSell {
+
+  // order_id is the unique ID of sell order.
+  uint64 order_id = 1;
+
+  // batch_denom is the credit batch being sold.
+  string batch_denom = 2;
+
+  // quantity is the quantity of credits being sold.
+  string quantity = 3;
+
+  // ask_price is the price the seller is asking for each unit of the
+  // batch_denom. Each credit unit of the batch will be sold for at least the
+  // ask_price or more.
+  cosmos.base.v1beta1.Coin ask_price = 4;
+
+  // disable_auto_retire disables auto-retirement of credits which allows a
+  // buyer to disable auto-retirement in their buy order enabling them to
+  // resell the credits to another buyer.
+  bool disable_auto_retire = 5;
+}
+
+// EventUpdateSellOrder is an event emitted when a sell order is updated.
+message EventUpdateSellOrder {
+
+  // owner is the owner of the sell orders.
+  string owner = 1;
+
+  //  sell_order_id is the ID of an existing sell order.
+  uint64 sell_order_id = 2;
+
+  // batch_denom is the credit batch being sold.
+  string batch_denom = 3;
+
+  // new_quantity is the updated quantity of credits available to sell, if it
+  // is set to zero then the order is cancelled.
+  string new_quantity = 4;
+
+  // new_ask_price is the new ask price for this sell order
+  cosmos.base.v1beta1.Coin new_ask_price = 5;
+
+  // disable_auto_retire updates the disable_auto_retire field in the sell order.
+  bool disable_auto_retire = 6;
+}
+
+// EventBuy is an event emitted when a buy order is created.
+message EventBuy {
```

Comment:
Should we have separate events when buy orders are created and filled?

Reply:
How would this work? We're not storing buy orders in state, so the created event would emit if partial or filled and the filled event would emit only if the buy order was filled?

Reply:
i think this would work for orders with filters, where buy orders can be made for a credit with X characteristics, but no direct sell order ID listed.. but sounds like it wont work for orders with direct SellOrderId's as with those you either fill it or you don't. 

Is there any spec on how the matching engine will work in this scheme for floating buy/sell orders?

Reply:
> how will the buy order be fulfilled at a later point in time?

Just like in regular exchanges, if the buy order is partially filled it will stay open and way for new or updated sell orders. Basically in every processing batch an attempt will be made to match it with sell orders.

We should probably have expiration times for both sell and buy orders btw.

Reply:
I think I misunderstood "filled". Yes, we should have a separate event for when a buy order is filled.

Reply:
At this point we havn't implemented any of the ORM tables for storing buy orders in state, but sounds like we would need this in order to actually process partial-filling correctly (where we keep a buy order open). If we don't implement that in this PR we could also just add it in a follow-up. When persisting the partially filled buy order to state, how would we store the `selection` ? Would it still reference the original `sell_order_id` ? Afaiu, the sell order would be exhausted and then closed (removed from state), so there wouldn't be any way for a seller to top up their supply on an existing sell order. They would rather create a new sell order.

If we allow for partial fills I'm not really sure how we can keep those buy orders open seeing that they are against sell order which would have already closed..

Reply:
If we address in a follow-up, it might make sense to include with #623 assuming that we would only need to store a buy order if it was using the filter option (not a direct buy order).

I added a second event for filling buy orders to distinguish from creating buy orders. At the moment the second event seems unnecessary given that only direct buy orders are implemented in this pull request but it makes sense when we implement the filter option. I think this goes in line with #623 as well.

I also added a couple commits to update the remaining quantity if the buy order does not purchase all credits within the sell order and to delete the sell order if there are no remaining credits.

Maybe we should break out a separate issue for adding and handling expiration times..?